### PR TITLE
[PIR] Refine pd_op_to_kernel_pass

### PR DESCRIPTION
**Pull request by zhangbo9674**
### PR types

New features
### PR changes

Others
### Description

Refine pd_op_to_kernel_pass
Pcard-67164

### CODE REVIEW

File: paddle/fluid/pir/transforms/pd_op_to_kernel_pass.cc [hunk 1]
```diff
@@ -146,14 +113,31 @@ bool NeedFallBackFromGPUDNN2GPU(pir::Operation* op,
       return true;
     }
   }
-
   return false;
 }
 
-std::set<std::string> GetSkipFeedNames(pir::Block* block) {
+static phi::Backend DeriveBackend(const std::string& op,
+                                  phi::Place place,
```
Comment (Aurelius84) [inline paddle/fluid/pir/transforms/pd_op_to_kernel_pass.cc]:
```suggestion
                                  cost phi::Place& place,
```

File: paddle/fluid/pir/transforms/pd_op_to_kernel_pass.cc [hunk 2]
```diff
@@ -163,191 +147,139 @@ std::set<std::string> GetSkipFeedNames(pir::Block* block) {
...
-        out_dtype,
-        selected_rows_type.dims(),
-        selected_rows_type.data_layout(),
-        selected_rows_type.lod(),
-        selected_rows_type.offset());
-  } else {
-    PADDLE_THROW(phi::errors::Unimplemented(
-        "BuildOutputType only support DenseTensorType and SelectedRowsType"));
-  }
+template <class IrT>
```
Comment (Aurelius84) [inline paddle/fluid/pir/transforms/pd_op_to_kernel_pass.cc]:
```suggestion
template <class IrType>
```
可以考虑跟前面保持一致

File: paddle/fluid/pir/transforms/pd_op_to_kernel_pass.cc [hunk 3]
```diff
@@ -417,54 +338,73 @@ pir::OpResult AddDtypeTransferOp(pir::Value in,
...
-    return dialect::AllocatedSelectedRowsType::get(
-        ctx,
-        place,
-        out_dtype,
-        selected_rows_type.dims(),
-        selected_rows_type.data_layout(),
-        selected_rows_type.lod(),
-        selected_rows_type.offset());
+
+template <class T>
```
Comment (Aurelius84) [inline paddle/fluid/pir/transforms/pd_op_to_kernel_pass.cc]:
```suggestion
template <class IrType>
```
这里看起来也是IrType? 可以统一下，可读性更好一些？

File: paddle/fluid/pir/transforms/pd_op_to_kernel_pass.cc [hunk 4]
```diff
@@ -417,54 +338,73 @@ pir::OpResult AddDtypeTransferOp(pir::Value in,
...
+    return create_allocated_type<AllocatedSelectedRowsType>(
+        type, place, data_dtype, ctx);
   } else {
     PADDLE_THROW(phi::errors::Unimplemented(
         "BuildOutputType only support DenseTensorType and SelectedRowsType"));
   }
 }
 
-phi::DataType GetKernelDataTypeByYamlInfo(
+template <class IrT1, class IrT2>
```
Comment (Aurelius84) [inline paddle/fluid/pir/transforms/pd_op_to_kernel_pass.cc]: 同上

File: paddle/fluid/pir/transforms/pd_op_to_kernel_pass.cc (hunk 1 above)
Comment (zhangbo9674) [reply paddle/fluid/pir/transforms/pd_op_to_kernel_pass.cc]: done, thanks!

File: paddle/fluid/pir/transforms/pd_op_to_kernel_pass.cc (hunk 2 above)
Comment (zhangbo9674) [reply paddle/fluid/pir/transforms/pd_op_to_kernel_pass.cc]: done, thanks!

File: paddle/fluid/pir/transforms/pd_op_to_kernel_pass.cc (hunk 3 above)
Comment (zhangbo9674) [reply paddle/fluid/pir/transforms/pd_op_to_kernel_pass.cc]: done, thanks!

**Review by Aurelius84 — APPROVED** [review]

**Pull request merged**